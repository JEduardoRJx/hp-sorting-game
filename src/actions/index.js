export const setCharacters = characters => ({
  type: 'SET_CHAR',
  characters
})

export const setGameCharacters = allCharacters => ({
  type: 'SET_GAME_CHAR',
  allCharacters
})

export const isLoading = bool => ({
  type: 'IS_LOADING',
  bool
})

export const setHouse = house => ({
  type: 'SET_HOUSE',
  house
})