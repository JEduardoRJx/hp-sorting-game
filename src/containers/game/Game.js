import React, { Component} from 'react';
import { connect } from 'react-redux';
import './Game.scss'
import GameChar from '../../components/gamechar/GameChar'

class Game extends Component {
  constructor() {
    super();
    this.state ={
      index: 0,
      correct: 0, 
      message: '',
      toggleModal: 'hide'
    }
  }

  checkAnswer = (event) => {
    console.log(event)
    const { gameCharacters } = this.props
    if (event.target.name === gameCharacters[this.state.index].house) {
      this.setState({index: this.state.index + 1, correct: this.state.correct + 1, message: 'Correct!', toggleModal: 'show'})
      setTimeout(() => {
        this.setState({toggleModal: 'hide'})
      }, 1500);
    } else {
      this.setState({message: 'Incorrect', toggleModal: 'show'});
      setTimeout(() => {
        this.setState({toggleModal: 'hide'})
      }, 1500);
    }
    this.setState({index: this.state.index + 1})
  }

  render() {
    const { gameCharacters, isLoading } = this.props;
    const message = <div className={`game-modal ${this.state.toggleModal}`}>
        <h1 className='game-modal-h1'>{this.state.message}</h1>
      </div>

    return (
      <section className='game-section'>
        {message} 
        {this.state.index <= 9 && <GameChar character={gameCharacters[this.state.index]}
          isLoading={isLoading} 
          checkAnswer={this.checkAnswer}/>}
        {this.state.index === 10 && <p>results</p>}
      </section>
    )
  }
}

const mapStateToProps = ({gameCharacters, isLoading}) => ({
  gameCharacters,
  isLoading
})

export default connect(mapStateToProps)(Game);