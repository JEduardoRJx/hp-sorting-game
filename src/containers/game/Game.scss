$hw-dark: #4c3638;
$hw-light: #D3A481;
$sl-dark: #276F45;
$sl-light: #A8B0BF;
$gryf-dark: #702737;
$gryf-light: #DAA520;
$rav-dark: #274F70;
$rav-light: #C98C4E;
$huff-dark: #141414;
$huff-light: #D1B738;

.game-section {
  height: 90vh;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
}

.game-results {
  background-color: $hw-light;
  grid-column: 2/4;
  padding: 25px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
  color: $hw-dark;
}

.game-results-title {
  font-size: 3em;
}

.game-results-score {
  font-size: 2em;
}

.game-results-quote {
  font-size: 2em;
}

.home-btn {
  font-size: 2em;
  text-decoration: none;
  background-color: $hw-light;
  color: $hw-dark;
  border-radius: 3px;
  padding: 5px;
  border: 3px solid $hw-dark;
  &:hover {
    background-color: $hw-dark;
    color: $hw-light;
  }
}