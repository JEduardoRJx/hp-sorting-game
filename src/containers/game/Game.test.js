import React from 'react';
import { shallow } from 'enzyme';
import { Game, mapStateToProps, mapDispatchToProps } from './Game'
import { setCharacters, isLoading, setGameCharacters } from '../../actions';

describe('GameContainer', () => {

  describe('Game component', () => {
    let wrapper;
    const mockSetCharacters = jest.fn();
    const mockLoading = jest.fn();
    const mockSetGameCharacters = jest.fn();
    const mockAllCharacters = [{
      bloodStatus: "pure-blood",
      deathEater: false,
      dumbledoresArmy: false,
      house: "Hufflepuff",
      ministryOfMagic: false,
      name: "Cedric Diggory",
      orderOfThePhoenix: false,
      role: "Student",
      school: "Hogwarts School of Witchcraft, and Wizardry",
      species: "human",
      wand: 'Ash, 12 1/4", unicorn hair',
      __v: 0,
      _id: "5a1096253dc2080021cd875f"
    }];
    const mockIsLoading = true;
    const mockGameCharacters = [{
      bloodStatus: "pure-blood",
      deathEater: false,
      dumbledoresArmy: false,
      house: "Hufflepuff",
      ministryOfMagic: false,
      name: "Cedric Diggory",
      orderOfThePhoenix: false,
      role: "Student",
      school: "Hogwarts School of Witchcraft, and Wizardry",
      species: "human",
      wand: 'Ash, 12 1/4", unicorn hair',
      __v: 0,
      _id: "5a1096253dc2080021cd875f"
    }];
    const mockUser = {
      name: 'name',
      wizardkind: 'wizard',
      wand: 'wand',
      house: 'slytherin'
    };

    beforeEach(() => {
      wrapper = shallow(<Game 
        setCharacters={mockSetCharacters}
        loading={mockLoading}
        setGameCharacters={mockSetGameCharacters}
        characters={mockAllCharacters}
        gameCharacters={mockGameCharacters}
        isLoading={mockIsLoading}
        user={mockUser}
      />);  
    });
    
    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should be able to check answer', () => {
      const mockEvent = {
        target: {
          name: 'name'
        }
      }
      wrapper.instance().checkAnswer(mockEvent);
      expect(wrapper.state('index')).toEqual(1);
      expect(wrapper.state('correct')).toEqual(0);
      expect(wrapper.state('message')).toEqual('Try Again!');
      expect(wrapper.state('toggleModal')).toEqual('show');
    })
  });

  describe('mapStateToProps', () => {
    it('should return a gameCharacters array', () => {
      const mockState = {
        filter: 'SET_CHAR',
        gameCharacters: [{
          bloodStatus: "pure-blood",
          deathEater: false,
          dumbledoresArmy: false,
          house: "Hufflepuff",
          ministryOfMagic: false,
          name: "Cedric Diggory",
          orderOfThePhoenix: false,
          role: "Student",
          school: "Hogwarts School of Witchcraft, and Wizardry",
          species: "human",
          wand: 'Ash, 12 1/4", unicorn hair',
          __v: 0,
          _id: "5a1096253dc2080021cd875f"
        }]
      }
      const expected = {
        gameCharacters: [{
          bloodStatus: "pure-blood",
          deathEater: false,
          dumbledoresArmy: false,
          house: "Hufflepuff",
          ministryOfMagic: false,
          name: "Cedric Diggory",
          orderOfThePhoenix: false,
          role: "Student",
          school: "Hogwarts School of Witchcraft, and Wizardry",
          species: "human",
          wand: 'Ash, 12 1/4", unicorn hair',
          __v: 0,
          _id: "5a1096253dc2080021cd875f"
        }]
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });

    it('should return isLoading', () => {
      const mockState = {
        filter: 'IS_LOADING',
        isLoading: false
      }
      const expected = {
        isLoading: false
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected)
    });

    it('should return an allCharacters array', () => {
      const mockState = {
        filter: 'SET_CHAR',
        allCharacters: [{
          bloodStatus: "pure-blood",
          deathEater: false,
          dumbledoresArmy: false,
          house: "Hufflepuff",
          ministryOfMagic: false,
          name: "Cedric Diggory",
          orderOfThePhoenix: false,
          role: "Student",
          school: "Hogwarts School of Witchcraft, and Wizardry",
          species: "human",
          wand: 'Ash, 12 1/4", unicorn hair',
          __v: 0,
          _id: "5a1096253dc2080021cd875f"
        }]
      }
      const expected = {
        allCharacters: [{
          bloodStatus: "pure-blood",
          deathEater: false,
          dumbledoresArmy: false,
          house: "Hufflepuff",
          ministryOfMagic: false,
          name: "Cedric Diggory",
          orderOfThePhoenix: false,
          role: "Student",
          school: "Hogwarts School of Witchcraft, and Wizardry",
          species: "human",
          wand: 'Ash, 12 1/4", unicorn hair',
          __v: 0,
          _id: "5a1096253dc2080021cd875f"
        }]
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });

    it('should return user', () => {
      const mockState = {
        filter: 'SET_USER',
        user: {
          name: 'name',
          wizardkind: 'wizard',
          wand: 'wand',
          house: 'slytherin'
        }
      }
      const expected = {
        user: {
          name: 'name',
          wizardkind: 'wizard',
          wand: 'wand',
          house: 'slytherin'
        }
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('calls dispatch with characters when setCharacters is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setCharacters([{
        bloodStatus: "pure-blood",
        deathEater: false,
        dumbledoresArmy: false,
        house: "Hufflepuff",
        ministryOfMagic: false,
        name: "Cedric Diggory",
        orderOfThePhoenix: false,
        role: "Student",
        school: "Hogwarts School of Witchcraft, and Wizardry",
        species: "human",
        wand: 'Ash, 12 1/4", unicorn hair',
        __v: 0,
        _id: "5a1096253dc2080021cd875f"
      }]);

      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.setCharacters([{
        bloodStatus: "pure-blood",
        deathEater: false,
        dumbledoresArmy: false,
        house: "Hufflepuff",
        ministryOfMagic: false,
        name: "Cedric Diggory",
        orderOfThePhoenix: false,
        role: "Student",
        school: "Hogwarts School of Witchcraft, and Wizardry",
        species: "human",
        wand: 'Ash, 12 1/4", unicorn hair',
        __v: 0,
        _id: "5a1096253dc2080021cd875f"
      }]);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with boolean when isLoading is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = isLoading(true);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.loading(true);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('calls dispatch with allCharacters when gameCharacters is called', () => {
      const mockDispatch = jest.fn();
      const mockAllCharacters = [{
        bloodStatus: "pure-blood",
        deathEater: false,
        dumbledoresArmy: false,
        house: "Hufflepuff",
        ministryOfMagic: false,
        name: "Cedric Diggory",
        orderOfThePhoenix: false,
        role: "Student",
        school: "Hogwarts School of Witchcraft, and Wizardry",
        species: "human",
        wand: 'Ash, 12 1/4", unicorn hair',
        __v: 0,
        _id: "5a1096253dc2080021cd875f"
      }]
      const actionToDispatch = setGameCharacters(mockAllCharacters);
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setGameCharacters(mockAllCharacters);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  })
})