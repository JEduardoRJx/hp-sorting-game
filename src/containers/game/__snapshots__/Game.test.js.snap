// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameContainer Game component should match snapshot 1`] = `
<section
  className="game-section"
>
  <Connect(PlayerInfo) />
  <Connect(GameChar)
    character={
      Object {
        "__v": 0,
        "_id": "5a1096253dc2080021cd875f",
        "bloodStatus": "pure-blood",
        "deathEater": false,
        "dumbledoresArmy": false,
        "house": "Hufflepuff",
        "ministryOfMagic": false,
        "name": "Cedric Diggory",
        "orderOfThePhoenix": false,
        "role": "Student",
        "school": "Hogwarts School of Witchcraft, and Wizardry",
        "species": "human",
        "wand": "Ash, 12 1/4\\", unicorn hair",
      }
    }
    checkAnswer={[Function]}
    isLoading={true}
    message=""
    toggleModal="hide"
  />
</section>
`;
